source/extensions/filters/network/thrift_proxy/header_transport_impl.cc

```diff
@@ -0,0 +1,299 @@
+namespace ThriftProxy {
+namespace {
+
+// c.f.
+// https://github.com/apache/thrift/blob/master/lib/cpp/src/thrift/protocol/TProtocolTypes.h#L27
+enum class HeaderProtocolType {
+  Binary = 0,
+  JSON = 1,
+  Compact = 2,
+
+  FirstHeaderProtocolType = Binary,
+  LastHeaderProtocolType = Compact,
+};
+
+} // namespace
+
+bool HeaderTransportImpl::decodeFrameStart(Buffer::Instance& buffer, MessageMetadata& metadata) {
+  if (buffer.length() < 14) {
+    return false;
+  }
+
+  // Size of frame, not including the length bytes.
+  const int32_t frame_size = BufferHelper::peekI32(buffer);
+
+  // Minimum header frame size is 18 bytes (4 bytes of frame size + 10 bytes of fixed header +
+  // minimum 4 bytes of variable header data), so frame_size must be at least 14.
+  if (frame_size < 14 || frame_size > MaxFrameSize) {
+    throw EnvoyException(fmt::format("invalid thrift header transport frame size {}", frame_size));
+  }
+
+  int16_t magic = BufferHelper::peekU16(buffer, 4);
+  if (!isMagic(magic)) {
+    throw EnvoyException(fmt::format("invalid thrift header transport magic {:04x}", magic));
+  }
+
+  // offset 6: 16 bit flags field, unused
+  // offset 8: 32 bit sequence number field
+  int32_t seq_id = BufferHelper::peekI32(buffer, 8);
+
+  // offset 12: 16 bit (remaining) header size / 4 (spec erroneously claims / 32).
+  int16_t raw_header_size = BufferHelper::peekI16(buffer, 12);
+  int32_t header_size = static_cast<int32_t>(raw_header_size) * 4;
+  if (header_size < 0 || header_size > MaxHeadersSize) {
+    throw EnvoyException(fmt::format("invalid thrift header transport header size {} ({:04x})",
+                                     header_size, static_cast<uint16_t>(raw_header_size)));
+  }
+
+  if (header_size == 0) {
+    throw EnvoyException("no header data");
+  }
+
+  if (buffer.length() < static_cast<uint64_t>(header_size) + 14) {
+    // Need more header data.
+    return false;
+  }
+
+  // Header data starts at offset 14 (4 bytes of frame size followed by 10 bytes of fixed header).
+  buffer.drain(14);
+
+  // Remaining frame size is the original frame size (which does not count itself), less the 10
+  // fixed bytes of the header (magic, flags, etc), less the size of the variable header data
+  // (header_size).
+  metadata.setFrameSize(static_cast<uint32_t>(frame_size - header_size - 10));
+  metadata.setSequenceId(seq_id);
+
+  ProtocolType proto = ProtocolType::Auto;
+  HeaderProtocolType header_proto =
+      static_cast<HeaderProtocolType>(drainVarIntI16(buffer, header_size, "protocol id"));
+  switch (header_proto) {
+  case HeaderProtocolType::Binary:
+    proto = ProtocolType::Binary;
+    break;
+  case HeaderProtocolType::Compact:
+    proto = ProtocolType::Compact;
+    break;
+  default:
+    throw EnvoyException(fmt::format("Unknown protocol {}", static_cast<int>(header_proto)));
+  }
+  metadata.setProtocol(proto);
+
+  int16_t num_xforms = drainVarIntI16(buffer, header_size, "transform count");
```

Comment:
Is this line vulnerable to a buffer read overflow? Here's the scenario I'm worried about:
* A malicious client sends a 15-byte message where `header_size` = 1
* This message gets through all the checks on lines 57-66 above.
* Then line 72 removes the first 14 bytes, leaving 1 byte in `buffer`.
* Depending on the value in that 1 remaining byte, `drainVarInt16` may try to read the byte after it, which is beyond the end of `buffer`.

Reply:
I don't believe so. We require that the full header (as given by header size) by available. As we read bytes from the header we decrement header_size (which is passed as a reference to drainVarIntI16). drainVarIntI16 is just drainVarIntI32 with an upper bound check. The latter requires header_size to be at least 1 and then invokes BufferHelper::peekVarIntI32 which validates the buffer has enough data before reading each byte that composes the value. If the peek reports a buffer underflow (buffer doesn't have enough data to read the whole number) or requires more bytes than are left in the header (supposing the buffer has more data beyond the end of the header) we throw.

That said, I'll add a test that exercises this directly. Mostly as a remdinder to myself: It's basically [this one](https://github.com/envoyproxy/envoy/pull/4082/files#diff-464b7abb89897074c8d6d7964d2cddd3R160) except that the number of transforms should be  truncated.